README.md
NBAMSeq
===========

`NBAMSeq` is a Bioconductor package for differential expression analysis based on negative binomial additive model. 

The Bioconductor package can be found [here](https://bioconductor.org/packages/release/bioc/html/NBAMSeq.html).    
The Bioconductor package vignette can be found [here](https://bioconductor.org/packages/release/bioc/vignettes/NBAMSeq/inst/doc/NBAMSeq-vignette.html).   
The `NBAMSeq` paper can be found [here](https://doi.org/10.1186/s12859-020-3506-x).

Package installation
------------
```{r}    
if (!requireNamespace("BiocManager", quietly = TRUE))  install.packages("BiocManager")    
BiocManager::install("NBAMSeq")    
```

Citation
------------
Ren, X., & Kuan, P. F. (2020). Negative binomial additive model for RNA-Seq data analysis. BMC Bioinformatics, 21(1), 1-15.

@article{ren2020negative,    
title={Negative binomial additive model for RNA-Seq data analysis},    
author={Ren, Xu and Kuan, Pei-Fen},    
journal={BMC Bioinformatics},    
volume={21},    
number={1},    
pages={1--15},    
year={2020},    
publisher={BioMed Central}    
}


inst/NEWS
Changes in version 1.5.1 (2020-05-04)
+ Updated citation

Changes in version 1.3.1 (2020-02-18)
+ The default gamma parameter in NBAMSeq changed from 2 to 2.5.
+ The NBAMSeq function now supports fitting linear model after 
  fitting GAM model.

Changes in version 1.1.1 (2019-08-16)
+ Added a new function makeplot.
+ Visualization part in vignette is modified.

Changes in version 0.99.1 (2019-02-13)
+ Package created.


tests/testthat.R
library(testthat)
library(NBAMSeq)

test_check("NBAMSeq")


tests/testthat/test_NBAMSeq.R
context("Test NBAMSeq")

test_that("Test invalid input", {

    nrows = 5
    ncols = 10
    counts = matrix(rnbinom(nrows * ncols, 10, 0.3), nrows)
    pheno = runif(ncols, 20, 80)
    colData = DataFrame(pheno = pheno)
    gsd = NBAMSeqDataSet(countData = counts,
        colData = colData, design = ~s(pheno))
    expect_error(NBAMSeq(gsd, gamma = "unknown"))
    expect_error(NBAMSeq(gsd, gamma = -2),
                 " 'gamma' should be greater or equal to 1.")
    expect_error(NBAMSeq(gsd, gamma = c(1,2)))
    expect_error(NBAMSeq(gsd, parallel = 1))
    expect_error(NBAMSeq(gsd, parallel = c(TRUE, FALSE)))
    expect_error(NBAMSeq(gsd, fitlin = 1))
    expect_error(NBAMSeq(gsd, fitlin = c(TRUE, FALSE)))

    #dds = makeExampleDESeqDataSet()
    #expect_error(NBAMSeq(dds))

})


test_that("Test NBAMSeq output", {
    # gsd = makeExample(n=5,m=10)
    ## Test parallel option
    # gsd = NBAMSeq(gsd, parallel = TRUE)
    # gsd2 = NBAMSeq(gsd, parallel = FALSE)
    # expect_identical(gsd,gsd2)
    
    #expect_true("Intercept"%in%names(mcols(gsd)))
    #expect_true("edf_pheno"%in%names(mcols(gsd)))
    #expect_true("Chisq_pheno"%in%names(mcols(gsd)))
    #expect_true("PValue_pheno"%in%names(mcols(gsd)))
    #expect_true("df_residual"%in%names(mcols(gsd)))
    #expect_true("null_deviance"%in%names(mcols(gsd)))
    #expect_true("df_null"%in%names(mcols(gsd)))

    n = 3
    m = 30
    pheno = runif(m, 20, 80)
    mu = matrix(rep(NA, n*m), nrow = n)
    interceptMean = 3
    interceptSD = 2
    betamat = matrix(rep(NA, n*3), nrow = n)
    betamat[,1] = rnorm(n, 3, 2)
    betamat[,2] = 0.2
    betamat[,3] = -0.0025

    x = cbind(rep(1,m), pheno, pheno^2)
    mu = t(2^(x %*% t(betamat))) + 100
    mumean = apply(mu, 1, mean)
    disp = 10/mumean + 0.001

    countData = matrix(rnbinom(m*n, mu=mu, size=1/disp), ncol=m)
    mode(countData) = "integer"
    colnames(countData) = paste0("sample", 1:m)
    rownames(countData) = paste0("gene", 1:n)

    var1 = rnorm(m)
    var2 = rnorm(m)
    var3 = rnorm(m)
    var4 = as.factor(sample(c(0,1,2), m, replace = TRUE))
    colData = data.frame(pheno = pheno, var1 = var1, var2 = var2,
                         var3 = var3, var4 = var4)
    gsd = NBAMSeqDataSet(countData = countData, colData = colData,
                        design = ~ var4 + s(pheno) + s(var1) + var2 + var3)

    ## Test getsf and setsf
    sf = getsf(gsd)
    expect_true(is.null(sf))
    
    sf = sample(1:5,m, replace = TRUE)
    setsf(gsd) = sf
    expect_true(all(getsf(gsd) == sf))
    
    # #### The following code are comments in order to pass the Bioconductor
    # ## check 5 min limit
    # 
    # gsd = NBAMSeqDataSet(countData = countData, colData = colData,
    #                      design = ~ var4 + s(pheno) + s(var1) + var2 + var3)
    # 
    # gsd = NBAMSeq(gsd, parallel = TRUE)
    # gsd2 = NBAMSeq(gsd, parallel = FALSE)
    # expect_identical(gsd,gsd2)
    # 
    # #check if gamma is working well
    # expect_true(all(mcols(gsd)$gamma == 2.5))
    # 
    # sf = getsf(gsd)
    # expect_true(is.numeric(sf))
    # expect_true(length(sf) == m)
    # expect_true(all(names(sf) == paste0("sample", 1:m)))
    # expect_true(all(getsf(gsd) == sf))
    # 
    # dds = DESeqDataSetFromMatrix(countData = assay(gsd),
    #                              colData = colData(gsd), design = ~pheno)
    # dds = estimateSizeFactors(dds)
    # #expect_true(all(sizeFactors(dds)==sf))
    # 
    # 
    # expect_true("Intercept"%in%names(mcols(gsd)))
    # expect_true("edf_var1"%in%names(mcols(gsd)))
    # expect_true("Chisq_var1"%in%names(mcols(gsd)))
    # expect_true("PValue_var1"%in%names(mcols(gsd)))
    # expect_true("var2"%in%names(mcols(gsd)))
    # expect_true("SE_var2"%in%names(mcols(gsd)))
    # expect_true("PValue_var2"%in%names(mcols(gsd)))
    # expect_true("smooth_pheno"%in%names(mcols(gsd)))
    # expect_true("smooth_var1"%in%names(mcols(gsd)))
    # expect_true("df_residual"%in%names(mcols(gsd)))
    # expect_true("null_deviance"%in%names(mcols(gsd)))
    # expect_true("df_null"%in%names(mcols(gsd)))
    # ## check whether AIC/BIC are saved in the object
    # expect_true("AIC"%in%names(mcols(gsd)))
    # expect_true("BIC"%in%names(mcols(gsd)))
    # 
    # # Test invalid input for results function
    # expect_error(results(gsd, name = 1))
    # expect_error(results(gsd, name = c("pheno","var1")))
    # expect_error(results(gsd),
    #              "Either 'name' or 'contrast' argument should be provided.")
    # expect_error(results(gsd, name = "var10"))
    # 
    # expect_error(results(gsd, name = "pheno", indepfilter = "unknown"))
    # expect_error(results(gsd, name = "pheno", indepfilter = c(TRUE, FALSE)))
    # 
    # expect_error(results(gsd, name = "pheno", alpha = "unknown"))
    # expect_error(results(gsd, name = "pheno", alpha = c(0.5, 0.6)))
    # expect_error(results(gsd, name = "pheno", alpha = -0.1))
    # expect_error(results(gsd, name = "pheno", alpha = 1.2))
    # 
    # 
    # ## test result function output
    # expect_error(results(gsd, name = "var4"),
    #              " 'name' should be a continuous variable. For factors,
    #             please use 'contrast' argument.")
    # expect_error(results(gsd, contrast = c("var4", 1)),
    #              " 'contrast' should be a character of length 3.")
    # expect_error(results(gsd, contrast = c("var4", 1, 1)),
    #              "2nd and 3rd element in constrast should be different.")
    # expect_error(results(gsd, contrast = c("var10", 1, 0)),
    #              "1st element in contrast should be a variable in colData.")
    # expect_error(results(gsd, contrast = c("var2", 1, 0)),
    #              "The variable in contrast should be a factor. For continuous
    #             variables, please use 'name' argument.")
    # expect_error(results(gsd, contrast = c("var4", 5, 0)),
    #              "2nd element in contrast should be an appropriate level.")
    # expect_error(results(gsd, contrast = c("var4", 1, 10)),
    #              "3rd element in contrast should be an appropriate level.")
    # 
    # res1 = results(gsd, name = "pheno")
    # expect_true("baseMean"%in%names(res1))
    # expect_true("edf"%in%names(res1))
    # expect_true("stat"%in%names(res1))
    # expect_true("pvalue"%in%names(res1))
    # expect_true("padj"%in%names(res1))
    # expect_true("AIC"%in%names(res1))
    # expect_true("BIC"%in%names(res1))
    # 
    # res2 = results(gsd, name = "var3")
    # expect_true("baseMean"%in%names(res2))
    # expect_true("coef"%in%names(res2))
    # expect_true("SE"%in%names(res2))
    # expect_true("stat"%in%names(res2))
    # expect_true("pvalue"%in%names(res2))
    # expect_true("padj"%in%names(res2))
    # expect_true("AIC"%in%names(res2))
    # expect_true("BIC"%in%names(res2))
    # 
    # ## The following code are comments in order to pass the Bioconductor
    # # ## check 5 min limit
    # ## check contrast
    # res3 = results(gsd, contrast = c("var4", 1, 2), parallel = TRUE)
    # res4 = results(gsd, contrast = c("var4", 2, 1), parallel = TRUE)
    # expect_identical(res3[["baseMean"]], res4[["baseMean"]])
    # expect_equal(res3[["coef"]],-res4[["coef"]], tolerance=1e-8)
    # expect_equal(res3[["SE"]],res4[["SE"]], tolerance=1e-8)
    # expect_equal(res3[["stat"]],-res4[["stat"]], tolerance=1e-8)
    # expect_equal(res3[["pvalue"]],res4[["pvalue"]], tolerance=1e-8)
    # expect_equal(res3[["padj"]],res4[["padj"]], tolerance=1e-8)
    # 
    # # ## check parallel option:
    # res7 = results(gsd, contrast = c("var4", 2, 1), parallel = FALSE)
    # expect_identical(res4[["baseMean"]], res7[["baseMean"]])
    # expect_equal(res4[["coef"]], res7[["coef"]], tolerance=1e-8)
    # expect_equal(res4[["SE"]], res7[["SE"]], tolerance=1e-8)
    # expect_equal(res4[["stat"]], res7[["stat"]], tolerance=1e-8)
    # expect_equal(res4[["pvalue"]], res7[["pvalue"]], tolerance=1e-8)
    # expect_equal(res4[["padj"]], res7[["padj"]], tolerance=1e-8)
    # 
    # 
    # res5 = results(gsd, contrast = c("var4", 1, 0))
    # res6 = results(gsd, contrast = c("var4", 0, 1))
    # expect_identical(res5[["baseMean"]], res6[["baseMean"]])
    # expect_identical(res5[["coef"]], -res6[["coef"]])
    # expect_identical(res5[["SE"]], res6[["SE"]])
    # expect_identical(res5[["stat"]], -res6[["stat"]])
    # expect_equal(res5[["pvalue"]],res6[["pvalue"]], tolerance=1e-8)
    # expect_equal(res5[["padj"]],res6[["padj"]], tolerance=1e-8)
    # 
    # expect_true(all(res5[["coef"]]==mcols(gsd)[["var4_1_vs_0"]]))
    # expect_true(all(res5[["SE"]]==mcols(gsd)[["SE_var4_1_vs_0"]]))
    # expect_true(all(res5[["pvalue"]]==mcols(gsd)[["PValue_var4_1_vs_0"]]))
    # 
    # gsd = NBAMSeqDataSet(countData = countData, colData = colData,
    #                      design = ~ var4 + s(pheno) + s(var1) + var2 + var3)
    # gsd1 = NBAMSeq(gsd, fitlin = TRUE, parallel = TRUE)
    # gsd2 = NBAMSeq(gsd, fitlin = TRUE, parallel = FALSE)
    # expect_identical(gsd1,gsd2)
    # expect_true(!"AIC"%in%names(mcols(gsd1)))
    # expect_true(!"BIC"%in%names(mcols(gsd1)))
    # expect_true("AICnonlin"%in%names(mcols(gsd1)))
    # expect_true("BICnonlin"%in%names(mcols(gsd1)))
    # expect_true("AICintercept"%in%names(mcols(gsd1)))
    # expect_true("BICintercept"%in%names(mcols(gsd1)))
    # expect_true("AIClin"%in%names(mcols(gsd1)))
    # expect_true("BIClin"%in%names(mcols(gsd1)))
    # 
    # res1 = results(gsd1, name = "pheno")
    # expect_true(!"AIC"%in%names(res1))
    # expect_true(!"BIC"%in%names(res1))
    # expect_true("AICnonlin"%in%names(res1))
    # expect_true("BICnonlin"%in%names(res1))
    # expect_true("AICintercept"%in%names(res1))
    # expect_true("BICintercept"%in%names(res1))
    # expect_true("BIClin"%in%names(res1))
    # expect_true("BIClin"%in%names(res1))
})






tests/testthat/test_methods.R

context("Test methods")

test_that("set setsf", {
    n = 5
    m = 10
    gsd = makeExample(n = n, m = m)
    sf = as.factor(sample(1:5,m, replace = TRUE))
    expect_error((setsf(gsd) = sf))
    sf = sample(1:5,m, replace = TRUE)
    sf[5] = NA
    expect_error((setsf(gsd) = sf))
    sf = sample(1:5,m-1, replace = TRUE)
    expect_error((setsf(gsd) = sf),
                 "Size factor length should be the same as number of samples.")
    sf = sample(1:5,m, replace = TRUE)
    sf[5] = -2
    expect_error((setsf(gsd) = sf), "Size factor cannot be 0 or negative.")

})





R/makeplot.R

#' @title Making plots to visualize nonlinear associations
#'
#' @description This function makes plots to visualize nonlinear associations.
#' @param object a NBAMSeqDataSet object
#' @param phenoname the name of nonlinear variable to be visualized
#' @param genename the name of gene to be visualized
#' @param ... additional arguments provided to \code{\link[mgcv]{plot.gam}}
#' @importFrom mgcv gam plot.gam
#' @return the plot made by plot.gam() function
#' @export
#' @examples
#' gsd = makeExample(n = 3, m = 10)
#' gsd = NBAMSeq(gsd)
#' makeplot(gsd, "pheno", "gene3", main = "gene10")


makeplot = function(object, phenoname, genename, ...){
    ## check input
    stopifnot(is(object, "NBAMSeqDataSet"))
    ## check whether NBAMSeq has been run before
    if(!metadata(object)$fitted){
        stop("NBAMSeq function should be run before making plots")
    }
    
    stopifnot(is.character(phenoname))
    stopifnot(length(phenoname)==1)
    stopifnot(phenoname%in%all.vars(getDesign(object)))
    ## check the variable is linear or nonlinear
    flag = grepl(paste0("s\\(",phenoname,"\\)"), 
        as.character(getDesign(object))[2])
    if(!flag){
        stop("Visulization can be made for nonlinear variables only")
    }
    
    stopifnot(is.character(genename))
    stopifnot(length(genename)==1)
    if(!genename%in%rownames(assay(object))){
        stop("genename is not valid")
    }
    ind = which(rownames(assay(object))==genename)
    
    formula_offset = update(getDesign(object), y ~ . + offset(logsf))
    gamDispMAP = mcols(object)$dispMAP
    dat = data.frame(colData(object))
    dat$y = assay(object)[ind,]
    gamfit = gam(formula_offset, family = nb(link = "log"), method = "REML", 
        gamma = mcols(object)[ind,"gamma"], data = dat)
    start = gamfit$coef
    ## get smoothing terms
    sterms = setdiff(attr(gamfit$terms,"term.labels"),
        attr(gamfit$pterms,"term.labels"))
    sps = unlist(mcols(object)[ind, paste0("smooth_",sterms)])
    
    gamFinalFit = gam(formula_offset, 
        family = negbin(theta = 1/gamDispMAP[ind],link = "log"),
        method = "REML", sp = sps, start = start, data = dat)
    
    varind = which(sterms==phenoname)
    plot.gam(gamFinalFit, select = varind, ...)
}





R/makeExample.R

#' @title Make an example NBAMSeqDataSet
#'
#' @description This function makes an example NBAMSeqDataSet
#' @param n number of genes
#' @param m number of samples
#' @importFrom stats rnbinom rnorm runif
#' @export
#' @return a NBAMSeqDataSet object
#' @references Love, M.I., Huber, W., Anders, S. (2014) Moderated estimation of
#' fold change and dispersion for RNA-seq data with DESeq2. Genome Biology,
#' 15:550. \url{https://doi.org/10.1186/s13059-014-0550-8}
#' @examples
#' gsd = makeExample()

makeExample <- function(n = 200, m = 30){
    pheno = runif(m, 20, 80)
    mu = matrix(rep(NA, n*m), nrow = n)
    interceptMean = 3
    interceptSD = 2
    betamat = matrix(rep(NA, n*3), nrow = n)
    betamat[,1] = rnorm(n, 3, 2)
    betamat[,2] = 0.2
    betamat[,3] = -0.0025

    x = cbind(rep(1,m), pheno, pheno^2)
    mu = t(2^(x %*% t(betamat))) + 100
    mumean = apply(mu, 1, mean)
    disp = 10/mumean + 0.001

    countData = matrix(rnbinom(m*n, mu=mu, size=1/disp), ncol=m)
    mode(countData) = "integer"
    colnames(countData) = paste0("sample", seq_len(m))
    rownames(countData) = paste0("gene", seq_len(n))

    colData = data.frame(pheno = pheno)
    NBAMSeqDataSet(countData = countData, colData = colData,design = ~ s(pheno))

}



R/results.R
#' @title Pulling out result
#'
#' @description This function pulls out result from NBAMSeqDataSet object
#' returned by \code{\link{NBAMSeq}}
#' @param object a NBAMSeqDataSet object returned by \code{\link{NBAMSeq}}
#' @param name the name of nonlinear variable or continuous linear variable
#' @param contrast a character of length 3. 1st element: name of factor
#' variable; 2nd element: name of numerator level; 3rd element: name of
#' denominator level. contrast = c("group", "treatment", "control") means
#' comparing treatment vs control for group variable.
#' @param indepfilter either TRUE or FALSE indicating whether independent
#' filtering should be performed. Default is TRUE.
#' @param alpha significant threhold for declaring genes as differentially
#' expressed. Default is 0.1.
#' @param pAdjustMethod pvalue adjustment method. Default is "BH". See
#' \code{\link[stats]{p.adjust}} for details.
#' @param parallel either TRUE or FALSE indicating whether parallel should be
#' used. Default is FALSE.
#' @param BPPARAM an argument provided to \code{\link{bplapply}}. See
#' \code{\link[BiocParallel]{register}} for details.
#' @param ... additional arguments provided to \code{pvalueAdjustment} function
#' in DESeq2. See \code{\link[DESeq2]{results}} for details.
#' @importFrom BiocParallel bplapply bpparam
#' @importFrom stats lowess p.adjust quantile
#' @importFrom genefilter filtered_p
#' @export
#' @return a DataFrame which contains the result
#' @references Love, M.I., Huber, W., Anders, S. (2014) Moderated estimation of
#' fold change and dispersion for RNA-seq data with DESeq2. Genome Biology,
#' 15:550. \url{https://doi.org/10.1186/s13059-014-0550-8}
#' @examples
#' gsd = makeExample(n = 3, m = 10)
#' gsd = NBAMSeq(gsd)
#' res = results(gsd, name = "pheno")

results <- function(object, name, contrast, indepfilter = TRUE, alpha = 0.1,
                    pAdjustMethod = "BH", parallel = FALSE,
                    BPPARAM = bpparam(), ...){

    ## check input
    stopifnot(is(object, "NBAMSeqDataSet"))
    ## check whether NBAMSeq has been run before
    if(!metadata(object)$fitted){
        stop("NBAMSeq function should be run before calling results function")
    }

    stopifnot(is.logical(indepfilter))
    stopifnot(length(indepfilter)==1)

    stopifnot(is.numeric(alpha))
    stopifnot(length(alpha)==1)
    stopifnot(alpha>0&alpha<1)

    if(missing(name)&missing(contrast)){
        stop("Either 'name' or 'contrast' argument should be provided.")
    } else if(!missing(name)&!missing(contrast)){
        stop(" 'name' and 'contrast' arguments cannot be both provided.")
    } else if(!missing(name)){
        stopifnot(is.character(name))
        stopifnot(length(name)==1)
        stopifnot(name%in%all.vars(getDesign(object)))

        ## check if name is a factor
        if(is.factor(colData(object)[[name]])){
            stop(" 'name' should be a continuous variable. For factors,
                please use 'contrast' argument.")
        }

        ## check the variable is linear or nonlinear
        flag = grepl(paste0("s\\(",name,"\\)"),
            as.character(getDesign(object))[2])
        if(flag){
            clnm = c("baseMean", paste0(c("edf_", "Chisq_", "PValue_"), name))
            resdf = mcols(object)[clnm]
            names(resdf) = c("baseMean", "edf", "stat", "pvalue")
        }else{
            clnm = c("baseMean", name, paste0(c("SE_", "PValue_"), name))
            resdf = mcols(object)[clnm]
            resdf$stat = resdf[[name]]/resdf[[paste0("SE_",name)]]
            names(resdf) = c("baseMean", "coef", "SE", "pvalue", "stat")
            resdf = resdf[c("baseMean", "coef", "SE", "stat", "pvalue")]
        }
    } else{
        if(!is.character(contrast)|length(contrast)!=3){
            stop(" 'contrast' should be a character of length 3.")
        }
        target = contrast[1]
        firstlevel = contrast[2]
        secondlevel = contrast[3]
        alllevels = levels(colData(object)[[target]])

        if(firstlevel==secondlevel){
            stop("2nd and 3rd element in constrast should be different.")
        }
        if(!target%in%names(colData(object))){
            stop("1st element in contrast should be a variable in colData.")
        }
        if(!is.factor(colData(object)[[target]])){
            stop("The variable in contrast should be a factor. For continuous
                variables, please use 'name' argument.")
        }
        if(!firstlevel%in%alllevels){
            stop("2nd element in contrast should be an appropriate level.")
        }
        if(!secondlevel%in%alllevels){
            stop("3rd element in contrast should be an appropriate level.")
        }

        baselevel = levels(colData(object)[[target]])[1]
        if(firstlevel!=baselevel&secondlevel!=baselevel){
            ## if both first level and second level are not base level,
            ## relevel the factor and refit the model
            dat = data.frame(colData(object))
            dat[,target] = factor(dat[,target],
                levels = c(secondlevel, setdiff(alllevels, secondlevel)))
            formula_offset = update(getDesign(object), y ~ . + offset(logsf))
            temp = paste0("smooth_", all.vars(getDesign(object)))
            smoothcol = temp[temp%in%names(mcols(object))]

            gamFit2 = function(i){
                dat$y = assay(object)[i,]   ## ith gene count
                gamFinalFit = gam(formula_offset,
                    family = negbin(theta = 1/mcols(object)[["dispMAP"]][3],
                        link = "log"), method = "REML",
                    sp = unlist(mcols(object)[smoothcol][i,]), data = dat)
                summary(gamFinalFit)$p.table[paste0(target, firstlevel),]
            }

            if(parallel){
                reslist = bplapply(seq_len(nrow(object)), gamFit2,
                    BPPARAM = BPPARAM)
            } else{
                reslist = lapply(seq_len(nrow(object)), gamFit2)
            }

            resdf = DataFrame(matrix(unlist(reslist), ncol = 4, byrow = TRUE))
            names(resdf) = c("coef", "SE", "stat", "pvalue")
            resdf = cbind(mcols(object)["baseMean"], resdf)
        }else{
            if(secondlevel==baselevel){
                clnm = c("baseMean", paste0(c("", "SE_", "PValue_"),
                    paste0(target,"_", firstlevel,"_vs_",secondlevel)))
            }else{
                clnm = c("baseMean", paste0(c("", "SE_", "PValue_"),
                    paste0(target,"_", secondlevel,"_vs_",firstlevel)))

            }
            resdf = mcols(object)[clnm]
            names(resdf) = c("baseMean", "coef", "SE", "pvalue")
            resdf$stat = resdf[["coef"]]/resdf[["SE"]]
            resdf = resdf[c("baseMean", "coef", "SE", "stat", "pvalue")]
            if(firstlevel==baselevel){
                resdf[["coef"]] = -resdf[["coef"]]
                resdf[["stat"]] = -resdf[["stat"]]
            }
        }
    }

    ## perform independent filtering
    resdf = pvalueAdjustment(resdf, independentFiltering = indepfilter,
        alpha = alpha, pAdjustMethod = pAdjustMethod, ...)
    rownames(resdf) = rownames(assay(object))
    AICBICnms = c("AIC", "BIC", "AICnonlin", "AIClin", "AICintercept", 
                  "BICnonlin", "BIClin", "BICintercept")
    AICBICcolind = which(names(mcols(object))%in%AICBICnms)
    resdf = cbind(resdf, DataFrame(mcols(object)[,AICBICcolind]))
    resdf
}


##############################################################################
# The function pvalueAdjustment is taken from DESeq2 package.
# Author: Michael Love (michaelisaiahlove@gmail.com)
# Reference: Love, M.I., Huber, W., Anders, S. (2014) Moderated estimation of
# fold change and dispersion for RNA-seq data with DESeq2. Genome Biology,
# 15:550.
##############################################################################

pvalueAdjustment <- function(res, independentFiltering, filter,
    theta, alpha, pAdjustMethod) {
    # perform independent filtering
    if (independentFiltering) {
        if (missing(filter)) {
            filter <- res$baseMean
        }
        if (missing(theta)) {
            lowerQuantile <- mean(filter == 0)
            if(lowerQuantile < .95) upperQuantile<- .95 else upperQuantile<- 1
            theta <- seq(lowerQuantile, upperQuantile, length=50)
        }
        # do filtering using genefilter
        stopifnot(length(theta) > 1)
        stopifnot(length(filter) == nrow(res))
        filtPadj <- filtered_p(filter=filter, test=res$pvalue,
            theta=theta, method=pAdjustMethod)
        numRej  <- colSums(filtPadj < alpha, na.rm = TRUE)
        # prevent over-aggressive filtering when all genes are null,
        # by requiring the max number of rejections is above a fitted curve.
        # If the max number of rejection is not greater than 10, then don't
        # perform independent filtering at all.
        lo.fit <- lowess(numRej ~ theta, f=1/5)
        if (max(numRej) <= 10) {
            j <- 1
        } else {
            residual <- if (all(numRej==0)) {
                0
            } else {
                numRej[numRej > 0] - lo.fit$y[numRej > 0]
            }
            thresh <- max(lo.fit$y) - sqrt(mean(residual^2))
            j <- if (any(numRej > thresh)) {
                which(numRej > thresh)[1]
            } else {
                1
            }
        }
        # j <- which.max(numRej) # old method
        padj <- filtPadj[, j, drop=TRUE]
        cutoffs <- quantile(filter, theta)
        filterThreshold <- cutoffs[j]
        filterNumRej <- data.frame(theta=theta, numRej=numRej)
        filterTheta <- theta[j]

        metadata(res)[["filterThreshold"]] <- filterThreshold
        metadata(res)[["filterTheta"]] <- filterTheta
        metadata(res)[["filterNumRej"]] <- filterNumRej
        metadata(res)[["lo.fit"]] <- lo.fit
        metadata(res)[["alpha"]] <- alpha
    } else {
        # regular p-value adjustment
        # does not include those rows which were removed
        # by maximum Cook's distance
        padj <- p.adjust(res$pvalue,method=pAdjustMethod)
    }
    res$padj <- padj
    res
}



